File: app/src/main/java/com/bongorian/signa1/DeviceProfile.java
package com.bongorian.signa1;

import android.app.ActivityManager;
import android.content.Context;

/** Conservative starting budget, refined by supported camera modes and runtime render/heat feedback. */
final class DeviceProfile {
    final boolean constrained;
    DeviceProfile(Context context){ActivityManager manager=(ActivityManager)context.getSystemService(Context.ACTIVITY_SERVICE);ActivityManager.MemoryInfo memory=new ActivityManager.MemoryInfo();if(manager!=null)manager.getMemoryInfo(memory);constrained=manager!=null&&(manager.isLowRamDevice()||memory.totalMem>0&&memory.totalMem<5L*1024*1024*1024);}
    long photoPixels(){return constrained?1_000_000:2_073_600;}
    long videoPixels(){return constrained?921_600:2_073_600;}
}

File: app/src/main/kotlin/com/bongorian/signa1/DeviceProfile.kt
package com.bongorian.signa1

import android.app.ActivityManager
import android.content.Context

/** Starting limits only: camera capabilities and measured render/thermal feedback refine them. */
internal class DeviceProfile(context: Context) {
    val constrained: Boolean
    private val budget: Budget

    init {
        val manager = context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager?
        val memory = ActivityManager.MemoryInfo()
        manager?.getMemoryInfo(memory)
        budget =
            classify(
                manager?.isLowRamDevice == true,
                memory.totalMem,
                Runtime.getRuntime().availableProcessors(),
            )
        constrained = budget.constrained
    }

    fun photoPixels(): Long = budget.photoPixels

    fun videoPixels(): Long = budget.videoPixels

    data class Budget(val constrained: Boolean, val photoPixels: Long, val videoPixels: Long)

    companion object {
        /** CPU count is a conservative hint, not a CPU/GPU benchmark. Unknown RAM starts low. */
        fun classify(lowRam: Boolean, memoryBytes: Long, cores: Int): Budget {
            val gib = 1024L * 1024 * 1024
            return when {
                lowRam || memoryBytes <= 0 || memoryBytes < 5 * gib || cores <= 4 ->
                    Budget(true, 921600, 307200)
                memoryBytes < 8 * gib || cores < 8 -> Budget(false, 1440000, 921600)
                else -> Budget(false, 2073600, 2073600)
            }
        }
    }
}
